Stricter validate_token checks

kangalio opened the pull request:
Implements #1032. In essence, `serenity::client::validate_token` has been changed to fully parse the token's components and check if they are meaningful, to catch more errors.

Tested on one of my bot tokens; the extracted user ID and token creation timestamp are correct.

Question: should the function that extracts the bot user ID and token creation timestamp out of the token be made public?

kangalio committed:
Stricter validate_token checks

Zaitara commented on src/client/mod.rs:
```diff
@@ -941,45 +943,59 @@ impl Client {
+        parse_token(token.as_ref()).ok_or(Error::Client(ClientError::InvalidToken))?;
 
-    if token.is_empty() {
+    dbg!(user_id, generation_time);
```
```suggestion
```

kangalio replied:
Oh crap how did I forget to remove those :facepalm:

Flat commented on src/client/mod.rs:
```diff
@@ -941,45 +943,59 @@ impl Client {
+        parse_token(token.as_ref()).ok_or(Error::Client(ClientError::InvalidToken))?;
 
-    if token.is_empty() {
+    dbg!(user_id, generation_time);
```
I don't think there is value in leaving this debug line in.

kangalio replied:
Yes, it was an oversight on my part, sorry

Flat commented on src/client/mod.rs:
```diff
@@ -941,45 +943,59 @@ impl Client {
+    dbg!(user_id, generation_time);
+
+    // Check if timestamps are in a sensible range
+    if user_id.created_at().year() >= 2100 || generation_time.year() >= 2100 {
```
This seems arbitrary and creates a Serenity Y2.1K problem.

kangalio replied:
It is fairly arbitrary. Do you have a better suggestion? I think it's not a great idea to query the current time in that function and compare to that; that feels outside the scope of what a "validate_token" function should do; plus it could break depending on OS time configuration.

Flat replied:
I do not think this is a necessary step for validation, at the end of the day it gets sent to discord to be accepted or rejected, this would be a more cursory review to ensure the formatting is proper I think. I don't think we are concerned with someone crafting a token with a huge or small timestamp which can't be used anyway.

Zaitara commented on src/client/mod.rs:
```diff
@@ -941,45 +943,59 @@ impl Client {
+
+/// Verifies that the token adheres to the Discord token format and extracts the bot user ID and the
+/// token generation timestamp
+fn parse_token(token: &str) -> Option<(UserId, chrono::NaiveDateTime)> {
```
```suggestion
pub fn parse_token(token: impl AsRef<str>) -> Option<(UserId, chrono::NaiveDateTime)> {
let token = token.as_ref();
```

kangalio replied:
So you're suggesting to make that method public. In that case would it be good to make a designated struct?
```rust
pub struct TokenComponents {
    bot_user_id: UserId,
    token_creation_time: chrono::NaiveDateTime,
}
```
That way, callers of the function can make it explicit what the "UserId" and the "NaiveDateTime" actually represent.

Zaitara commented on src/client/mod.rs:
```diff
@@ -941,45 +943,59 @@ impl Client {
-/// assert!(validate_token("a.abcdef.b").is_ok());
-/// assert!(validate_token("a.abcde.b").is_err());
+/// // helpful to prevent typos
+/// assert!(validate_token("Njg4NzYwMjQxMzYzODc3ODg4.C_ikow.j3VupLBuE1QWZng3TMGH0z_UAwg").is_err());
```
It is not a real token right?

kangalio replied:
It's not, it's an adaption of the pseudo token which was there before (see the line above).

acdenisSK commented on src/client/mod.rs:
```diff
@@ -941,45 +943,59 @@ impl Client {
+        parse_token(token.as_ref()).ok_or(Error::Client(ClientError::InvalidToken))?;
 
-    if token.is_empty() {
+    dbg!(user_id, generation_time);
```
```suggestion
```

Flat commented on src/client/mod.rs:
```diff
@@ -37,6 +37,7 @@ use std::{
     task::{Context as FutContext, Poll},
 };
 
+use chrono::Datelike;
```
@acdenisSK was there a push to remove chrono or are we fine with using it over time still?

arqunis replied:
I do not recall migrating to the `time` crate over `chrono`. It should be fine to continue using it.

kangalio committed:
Remove old debug statement (oops)

kangalio committed:
Make parse_token function public

kangalio committed:
Remove sensible timespan check in validate_token

Previously it checked if the year is <=2100. People said it will create a Serenity Y2.1K problem. Tbh I don't think serenity will exist 80 years in the future but eh
